In [2]:
import numpy as np
import pandas as pd

# TASK - 1

**READING CSV OR EXCEL FILE**


1.Import the 'Salaries.csv' dataset from a specified URL into a Pandas DataFrame
named data.


In [3]:
sal_data_url="https://raw.githubusercontent.com/Apress/data-analysis-and-visualization-using-python/master/Ch07/Salaries.csv"
sal_data=pd.read_csv(sal_data_url)
sal_data.head()

,rank,discipline,phd,service,sex,salary
0,Prof,B,56,49,Male,186960
1,Prof,A,12,6,Male,93000
2,Prof,A,23,20,Male,110515
3,Prof,A,40,31,Male,131205
4,Prof,B,20,18,Male,104800


2.Import the 'Cars.csv' dataset from a local file path into a Pandas DataFrame named data.

In [4]:
cars_data=pd.read_excel("Cars.xlsx")
cars_data.head()

,HP,MPG,VOL,SP,WT
0,49,53.700681,89,104.185353,28.762059
1,55,50.013401,92,105.461264,30.466833
2,55,50.013401,92,105.461264,30.193597
3,70,45.696322,92,113.461264,30.632114
4,53,50.504232,92,104.461264,29.889149


**HEAD**

Display the first few rows of a DataFrame named df with a preceding label.

In [5]:
df=pd.read_excel("Cars.xlsx")
df.head()

,HP,MPG,VOL,SP,WT
0,49,53.700681,89,104.185353,28.762059
1,55,50.013401,92,105.461264,30.466833
2,55,50.013401,92,105.461264,30.193597
3,70,45.696322,92,113.461264,30.632114
4,53,50.504232,92,104.461264,29.889149


**TAIL:**

In [6]:
df.tail()

,HP,MPG,VOL,SP,WT
76,322,36.900000,50,169.598513,16.132947
77,238,19.197888,115,150.576579,37.923113
78,263,34.000000,50,151.598513,15.769625
79,295,19.833733,119,167.944460,39.423099
80,236,12.101263,107,139.840817,34.948615


**SHAPE:**

In [7]:
df.shape

(81, 5)

**DESCRIBE:**

In [8]:
df.describe()

,HP,MPG,VOL,SP,WT
count,81.000000,81.000000,81.000000,81.000000,81.000000
mean,117.469136,34.422076,98.765432,121.540272,32.412577
std,57.113502,9.131445,22.301497,14.181432,7.492813
min,49.000000,12.101263,50.000000,99.564907,15.712859
25%,84.000000,27.856252,89.000000,113.829145,29.591768
50%,100.000000,35.152727,101.000000,118.208698,32.734518
75%,140.000000,39.531633,113.000000,126.404312,37.392524
max,322.000000,53.700681,160.000000,169.598513,52.997752


# TASK - 2

**DICTIONARY:**


In [9]:
data={
    'Name': ['John', 'Jane', 'Babu', 'Peter', 'Leju'],
    'Age': [25, 30, 35, 40, 55],
    'City': ['New York', 'London', 'Paris', 'UK', 'Germany']
}

**Step -1**

Create a Pandas DataFrame named df using the provided dictionary data and then
display the dictionary.

In [10]:
df=pd.DataFrame(data)
df

,Name,Age,City
0,John,25,New York
1,Jane,30,London
2,Babu,35,Paris
3,Peter,40,UK
4,Leju,55,Germany


**Step-2**

Define a dictionary named data with specific key-value pairs, create a Pandas
DataFrame named df from this dictionary, and display the dictionary's contents with a
preceding label.


In [11]:
print("DataFrame Creation:")
print(data)

DataFrame Creation:
{'Name': ['John', 'Jane', 'Babu', 'Peter', 'Leju'], 'Age': [25, 30, 35, 40, 55], 'City': ['New York', 'London', 'Paris', 'UK', 'Germany']}


**Step-3**

Display the first five rows of a DataFrame named df with a preceding label.

In [12]:
df.head()

,Name,Age,City
0,John,25,New York
1,Jane,30,London
2,Babu,35,Paris
3,Peter,40,UK
4,Leju,55,Germany


**Step-4**

Display a concise summary of the DataFrame named df, including information about
its index, columns, data types, non-null values, and memory usage, with a preceding
label

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Name    5 non-null      object
 1   Age     5 non-null      int64 
 2   City    5 non-null      object
dtypes: int64(1), object(2)
memory usage: 252.0+ bytes


# TASK - 3

You are working with a dataset of employee salaries, stored in a pandas DataFrame
named data. The dataset includes the following columns: Department, EmployeeID,
Salary, YearsOfExperience, and Age. You need to perform the following tasks using the
groupby() function:






In [14]:
s_data=pd.read_csv("employee_salaries.csv")
s_data

,Department,EmployeeID,Salary,YearsOfExperience,Age
0,Marketing,1001,63483,9,49
1,Sales,1002,88555,7,46
2,Engineering,1003,57159,9,54
3,Sales,1004,75920,8,22
4,Sales,1005,107121,12,48
5,Finance,1006,109479,2,34
6,Engineering,1007,59457,1,24
7,Engineering,1008,106557,16,27
8,Engineering,1009,117189,5,29
9,Sales,1010,118953,3,48


1. Write the code to group the data by Department and calculate the mean salary
for each group. What is the result of this operation? What insights can you draw
from the mean salaries of each department?

In [17]:
dept_groupby=s_data.groupby('Department')
dept_groupby['Salary'].mean()

Department
Engineering    76457.500000
Finance        70509.700000
HR             79437.142857
Marketing      85255.230769
Sales          82034.800000
Name: Salary, dtype: float64

**Insights**

- Marketing has the highest average salary (₹85,255), indicating it might be a high-performance team with more seniors.

- Finance shows the lowest average (₹70,509), indicating it having more entry-level employees.

- Engineering lies in the middle (₹76,457), it may have a balanced mix of junior and senior roles.

2. How can you group by multiple columns and apply multiple aggregation
functions? Write the code for this operation and explain the output.


In [18]:
multi_group = s_data.groupby(['Department', 'YearsOfExperience']).agg({
    'Salary': ['mean', 'max', 'min'],
    'Age': ['mean', 'max']
})
multi_group

Salary                        Age    
                                        mean     max     min       mean max
Department  YearsOfExperience                                              
Engineering 1                   59457.000000   59457   59457  24.000000  24
            5                  117189.000000  117189  117189  29.000000  29
            7                   88984.000000   88984   88984  53.000000  53
            8                   80757.000000   80757   80757  58.000000  58
            9                   57159.000000   57159   57159  54.000000  54
            10                  49268.000000   49268   49268  44.000000  44
            15                  78316.500000  114065   42568  51.000000  53
            16                 106557.000000  106557  106557  27.000000  27
            17                  48571.000000   48571   48571  22.000000  22
Finance     1                   85758.000000   85758   85758  45.000000  45
            2                  109479.000000  109479  109479  34.000000  34
            5                   95804.000000  111211   80397  38.500000  53
            10                  73159.000000   73159   73159  55.000000  55
            12                  52666.000000   52666   52666  32.000000  32
            13                  49622.000000   53986   42695  41.000000  58
            20                  43561.000000   43561   43561  54.000000  54
HR          3                   72606.000000   72606   72606  33.000000  33
            4                  107563.000000  107563  107563  32.000000  32
            5                  101228.000000  101228  101228  47.000000  47
            6                  101858.000000  101858  101858  43.000000  43
            9                   63247.000000   63247   63247  23.000000  23
            15                  64300.000000   64300   64300  23.000000  23
            19                  45258.000000   45258   45258  45.000000  45
Marketing   1                   51534.000000   51534   51534  23.000000  23
            2                   79504.000000   79504   79504  26.000000  26
            3                  112409.000000  112409  112409  36.000000  36
            4                   91214.000000   91214   91214  23.000000  23
            7                   87626.666667   95591   79099  50.333333  59
            9                   63483.000000   63483   63483  49.000000  49
            11                  66854.000000   66854   66854  30.000000  30
            12                  92995.000000   92995   92995  30.000000  30
            15                 105697.000000  105697  105697  53.000000  53
            19                 103704.000000  103704  103704  53.000000  53
            20                  78044.000000   78044   78044  40.000000  40
Sales       3                   84322.500000  118953   49692  51.000000  54
            7                   88555.000000   88555   88555  46.000000  46
            8                   75920.000000   75920   75920  22.000000  22
            11                 102592.000000  102592  102592  25.000000  25
            12                  92890.500000  107121   78660  42.500000  48
            13                  80774.000000   80774   80774  27.000000  27
            14                  59040.500000   77065   41016  51.500000  58

**Insights**

- Salary - mean: Average salary for employees in that department with those years of experience.

- Salary - max: Maximum salary in that group.

- Salary - min: Minimum salary in that group.

- Age - mean: Average age of employees in that group.

- Age - max: Oldest age in that group.

3. How would you filter the grouped data based on a condition (e.g., mean salary
greater than $60,000)? Write the code to perform this filtering and describe the
results.


In [20]:
multi_group[multi_group[('Salary', 'mean')] > 60000]

Salary                        Age    
                                        mean     max     min       mean max
Department  YearsOfExperience                                              
Engineering 5                  117189.000000  117189  117189  29.000000  29
            7                   88984.000000   88984   88984  53.000000  53
            8                   80757.000000   80757   80757  58.000000  58
            15                  78316.500000  114065   42568  51.000000  53
            16                 106557.000000  106557  106557  27.000000  27
Finance     1                   85758.000000   85758   85758  45.000000  45
            2                  109479.000000  109479  109479  34.000000  34
            5                   95804.000000  111211   80397  38.500000  53
            10                  73159.000000   73159   73159  55.000000  55
HR          3                   72606.000000   72606   72606  33.000000  33
            4                  107563.000000  107563  107563  32.000000  32
            5                  101228.000000  101228  101228  47.000000  47
            6                  101858.000000  101858  101858  43.000000  43
            9                   63247.000000   63247   63247  23.000000  23
            15                  64300.000000   64300   64300  23.000000  23
Marketing   2                   79504.000000   79504   79504  26.000000  26
            3                  112409.000000  112409  112409  36.000000  36
            4                   91214.000000   91214   91214  23.000000  23
            7                   87626.666667   95591   79099  50.333333  59
            9                   63483.000000   63483   63483  49.000000  49
            11                  66854.000000   66854   66854  30.000000  30
            12                  92995.000000   92995   92995  30.000000  30
            15                 105697.000000  105697  105697  53.000000  53
            19                 103704.000000  103704  103704  53.000000  53
            20                  78044.000000   78044   78044  40.000000  40
Sales       3                   84322.500000  118953   49692  51.000000  54
            7                   88555.000000   88555   88555  46.000000  46
            8                   75920.000000   75920   75920  22.000000  22
            11                 102592.000000  102592  102592  25.000000  25
            12                  92890.500000  107121   78660  42.500000  48
            13                  80774.000000   80774   80774  27.000000  27

4. How can you define and apply a custom aggregation function to each group?
Write the code for this operation and explain what the results represent.


In [21]:
def salary_range(s):
    return s.max() - s.min()
s_data.groupby('Department')['Salary'].agg(salary_range)

Department
Engineering    74621
Finance        68516
HR             62305
Marketing      60875
Sales          77937
Name: Salary, dtype: int64